### PR TITLE
Feature/dick/anomaly score normalization

### PR DESCRIPTION
**Pull request by djdameln**
# Description

- This PR adds anomaly score and anomaly normalization. A substantial refactor was needed to make this functionality compatible with the TorchMetrics evaluation design.

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
#
- Adds `NormalizationCallback`, which normalizes the predicted scores and error maps based on the distribution of normal data in the training set.
- A test case was added for the normalization callback, which asserts that the (image-level) performance with and without normalization is the same.
- Normalization is repeated with numpy operations in the post-process method of the openvino inferencer.
- Adaptive threshold and training set statistics are now implemented as `TorchMetrics` classes, which de-clutters the base anomaly module and allows automatic saving and loading of the values in the model's state dict.
- F1 score is now computed by using TorchMetrics `F1` class configured with the computed threshold, instead of `OptimalF1`.
- Our custom `FeatureExtractor` was replaced by torchvision's `create_feature_extractor`. For this I had to bump up the torch and torchvision versions.

# Checklist:

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: requirements/requirements.txt [hunk 1]
```diff
@@ -1,8 +1,9 @@
 pillow==8.3.2
 nncf>=2.0.0
 pytorch-lightning==1.3.6
-torch==1.8.1
-torchvision==0.9.1
+-f https://download.pytorch.org/whl/torch_stable.html
+torch==1.10.0+cu111
+torchvision==0.11.0+cu111
```
Comment (samet-akcay) [inline requirements/requirements.txt:6]: As per our discussion, I guess we'll remove cu111. We might need to check with ote that they are fine with these version.

File: tests/core/callbacks/normalization_callback/test_normalization_callback.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def run_train_test(config):
+    model = get_model(config)
+    datamodule = get_datamodule(config)
+    callbacks = get_callbacks(config)
+    trainer = Trainer(**config.trainer, callbacks=callbacks)
+    trainer.fit(model=model, datamodule=datamodule)
+    results = trainer.test(model=model, datamodule=datamodule)
+    return results
```
Comment (samet-akcay) [inline tests/core/callbacks/normalization_callback/test_normalization_callback.py:35]: Perhaps, we could place this above the test function

File: anomalib/models/padim/model.py [hunk 3]
```diff
@@ -26,9 +26,9 @@
 from kornia import gaussian_blur2d
 from omegaconf import ListConfig
 from torch import Tensor, nn
+from torchvision.models.feature_extraction import create_feature_extractor
 
 from anomalib.core.model import AnomalyModule
-from anomalib.core.model.feature_extractor import FeatureExtractor
```
Comment (samet-akcay) [inline anomalib/models/padim/model.py:31]: Now that torchvision supports feature extraction, should we still need this? In the end, torchvision version is a lot more efficient than this. Maybe we could add a deprecation warning and gradually remove it from the repo.

**Review by samet-akcay — COMMENTED** [review]
Thanks a lot @djdameln! Looks really good, only minor stuff.

**Review by ashwinvaidya17 — APPROVED** [review]

File: anomalib/core/metrics/training_stats.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+"""Module that computes the parameters of the normal data distribution of the training set."""
+from typing import Optional, Tuple
+
+import torch
+from torch import Tensor
+from torchmetrics import Metric
+
+
+class TrainingStats(Metric):
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/core/metrics/training_stats.py:9]: `TrainingDistribution`, `AnomalyScoreDistribution`, `NormalDataDistribution`, `TrainingSetStatistics`?

[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/core/metrics/training_stats.py:9]: `TrainingDistribution` or `NormalDataDistribution` sound more descriptive to me. What do you think @djdameln  and @ashwinvaidya17

[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/core/metrics/training_stats.py:9]: Technically, the class could also be applied to the validation set or some collection of abnormal data. So maybe I would vote for `AnomalyScoreDistribution`, which covers all cases.

[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/core/metrics/training_stats.py:9]: Because essentially the class parameterizes the distribution of the anomaly scores of some collection of data. The source of this collection could be any dataset or subset, and the type of labels in the dataset could be normal, anomalous or mixed.

[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/core/metrics/training_stats.py:9]: yes, good point! In this case, I'd also vote for `AnomalyScoreDistribution`

File: anomalib/core/callbacks/__init__.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 from .compress import CompressModelCallback
 from .model_loader import LoadModelCallback
+from .normalization import NormalizationCallback
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/core/callbacks/__init__.py:12]: Yeah, I tried `AnomalyScoreNormalizationCallback` but found it too verbose. `ScoreNormalizationCallback` would be better, though maybe a bit vague. How about `OutputNormalizationCallback`, because it normalizes the outputs of the model?

File: anomalib/data/tiler.py [hunk 6]
```diff
@@ -196,11 +196,11 @@ def __init__(
         self.num_patches_w: int
 
     @staticmethod
-    def __validate_size_type(parameter) -> Tuple:
+    def __validate_size_type(parameter: Union[int, Sequence]) -> Tuple[int, ...]:
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/data/tiler.py:199]: I tried `Tuple[int, int]`, but mypy complained about it. It does accept `Tuple[int, ...]`. Not sure why, I agree that `Tuple[int, int]` would seem more reasonable

File: anomalib/models/padim/model.py (hunk 3 above)
Comment (djdameln) [reply anomalib/models/padim/model.py:31]: Yeah, I decided to leave it for now because it might still be needed somewhere. The deprecation warning would be a good temporary solution

File: anomalib/core/callbacks/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/core/callbacks/__init__.py]: I changed it to `OutputNormalizationCallback`. Let me know what you think

**Comment by djdameln** [comment]
Thanks for the review @samet-akcay. I addressed your comments and updated the PR

File: anomalib/core/model/anomaly_module.py [hunk 7]
```diff
@@ -41,21 +41,19 @@ def __init__(self, params: Union[DictConfig, ListConfig]):
         self.save_hyperparameters(params)
         self.loss: torch.Tensor
         self.callbacks: List[Callback]
-        self.register_buffer("threshold", torch.tensor(params.model.threshold.default))  # pylint: disable=not-callable
-        self.threshold: torch.Tensor
+
+        self.image_threshold = AdaptiveThreshold(self.hparams.model.threshold.image_default)
+        self.pixel_threshold = AdaptiveThreshold(self.hparams.model.threshold.pixel_default)
+
+        self.training_stats = AnomalyScoreDistribution()
```
Comment (samet-akcay) [inline anomalib/core/model/anomaly_module.py:48]: Would it be an idea to also rename `self.training_stats` to something like `self.training_distribution`

File: requirements/requirements.txt [hunk 8]
```diff
@@ -1,8 +1,9 @@
 pillow==8.3.2
 nncf>=2.0.0
 pytorch-lightning==1.3.6
-torch==1.8.1
-torchvision==0.9.1
+-f https://download.pytorch.org/whl/torch_stable.html
```
Comment (samet-akcay) [inline requirements/requirements.txt:4]: I think we need to remove this line too.

File: anomalib/core/callbacks/normalization.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+from typing import Any, Dict, Optional
+
+import pytorch_lightning as pl
+import torch
+from pytorch_lightning import Callback, Trainer
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from torch.distributions import LogNormal, Normal
+
+
+class OutputNormalizationCallback(Callback):
```
Comment (samet-akcay) [inline anomalib/core/callbacks/normalization.py:12]: How about the name of your PR? :) Like `AnomalyScoreNormalizationCallback` or `ScoreNormalizationCallback`. It doesn't matter if it's verbose. What's more important is its readability

**Review by samet-akcay — COMMENTED** [review]
Done with the second round. This will be super cool once merged.

**Comment by samet-akcay** [comment]
Maybe we could use version 0.2.3 for this one?

File: tests/core/callbacks/normalization_callback/test_normalization_callback.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def test_normalizer():
+    config = get_configurable_parameters(model_config_path="anomalib/models/padim/config.yaml")
```
Comment (ashwinvaidya17) [inline tests/core/callbacks/normalization_callback/test_normalization_callback.py:20]: Can you add `get_dataset_path()` to `config.dataset.path`? Without this, the CI is downloading the entire dataset again in the current directory.  Due to the permissions of the original files, the CI cannot remove the dataset once the test runs.
